{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import scipy.stats as stats\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "db = sqlite3.connect('./finalDupCompDatabase')\n",
    "cursor = db.cursor()\n",
    "nodeDivAge = {'Opisthokonta':1105/50,\n",
    "             'Bilateria':824/50,\n",
    "             'Chordata':684/50,\n",
    "             'Vertebrata':615/50,\n",
    "             'Gnathostomata':473/50,\n",
    "             'Euteleostomi':435/50,\n",
    "             'Sarcopterygii':413/50,\n",
    "             'Tetrapoda':352/50,\n",
    "             'Amniota':312/50,\n",
    "             'Mammalia':177/50,\n",
    "             'Theria':159/50,\n",
    "             'Eutheria':105/50,\n",
    "             'Boreoeutheria':96/50,\n",
    "             'Euarchontoglires':90/50,\n",
    "             'Primates':74/50,\n",
    "             'Haplorrhini':67/50,\n",
    "             'Simiiformes':43/50,\n",
    "             'Catarrhini':29.4/50,\n",
    "             'Hominoidea':20.2/50,\n",
    "             'Hominidae':16.8/50,\n",
    "             'Homininae':9.1/50,\n",
    "             'Homo sapiens':6.7/50}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# duplication node based\n",
    "cursor.execute('''SELECT id, \n",
    "                        genLen, \n",
    "                        cdsLen,\n",
    "                        evolRate, \n",
    "                        u_domains,\n",
    "                        gc,\n",
    "                        gc3,\n",
    "                        domains,\n",
    "                        specificity,\n",
    "                        transCount,\n",
    "                        intCount,\n",
    "                        intAvg,\n",
    "                        max_exp,\n",
    "                        ess,\n",
    "                        motif_number_1k,\n",
    "                        PPIs,\n",
    "                        intCov,\n",
    "                        mis_Z_score,\n",
    "                        pLI_score,\n",
    "                        loftool_percentile,\n",
    "                        s_het,\n",
    "                        Phi,\n",
    "                        RVIS,\n",
    "                        EvoTol,\n",
    "                        CAI,\n",
    "                        IntDisProp,\n",
    "                        dupCat_maj,\n",
    "                        youngestNode,\n",
    "                        oldestNode\n",
    "                    FROM\n",
    "                        gene_features\n",
    "                    WHERE (dupCat_maj == \"WGD\" OR dupCat_maj == \"SSD\")\n",
    "                     AND EvoTol != \"Not expressed above threshold in this ontology\"\n",
    "                     AND CAI != \"Error-check\"\n",
    "                     AND RVIS != \"NA\" \n",
    "                     AND pLI_score != \"NA\" ''')\n",
    "res = [list(x) for x in cursor.fetchall()]\n",
    "df = pd.DataFrame(res)\n",
    "df.columns = ['id','gen_len','clen','evolRate','uDoms','gc','gc3',\n",
    "              'doms','spec','trans_count','int_count','avg_int',\n",
    "              'max_exp','ess','motifs','PPIs','int_cov','missenseZ','pLI',\n",
    "              'loftool','shet','phi','rvis','EvoTol','CAI','IntDis','dup_type','minAge','maxAge']\n",
    "df['minAge'] = [nodeDivAge[x] for x in df['minAge']]\n",
    "df['maxAge'] = [nodeDivAge[x] for x in df['maxAge']]\n",
    "df['ess'] = -(df['ess'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#model selection and diagnostics\n",
    "modelDict = {}\n",
    "featureList = []\n",
    "\n",
    "for col in df.columns:\n",
    "    noLog, noBox= False,False\n",
    "    if col == 'id' or col == 'dup_type' or col == 'minAge' or col == 'maxAge':\n",
    "        continue\n",
    "    else:\n",
    "        print(col)\n",
    "        featureList.append(col)\n",
    "        formula = col + '~ maxAge + dup_type'\n",
    "        formula2 = col + '~ maxAge + dup_type + maxAge*dup_type'\n",
    "        formula3 = 'np.log10(' + col + ') ~ maxAge + dup_type'\n",
    "        formula4 = 'np.log10(' + col + ') ~ maxAge + dup_type + maxAge*dup_type'\n",
    "        formula5 = 'stats.boxcox('+col + ')[0] ~ maxAge + dup_type'\n",
    "        formula6 = 'stats.boxcox('+ col + ')[0] ~ maxAge + dup_type + maxAge*dup_type'\n",
    "\n",
    "#         if col == 'shet':\n",
    "#             data = df_shet\n",
    "#         else:\n",
    "#             \n",
    "        data = df\n",
    "        if min(df[col]) <= 0:\n",
    "            df[col] = [x + (min([y for y in df[col] if y > 0]))/2 for x in df[col]]\n",
    "        \n",
    "        fittedModel = smf.ols(formula = formula, data = data).fit()\n",
    "        fittedModel2 = smf.ols(formula = formula2, data = data).fit()\n",
    "        try:\n",
    "            fittedModel3 = smf.ols(formula = formula3, data=data).fit()\n",
    "            fittedModel4 = smf.ols(formula = formula4, data=data).fit()\n",
    "        except:\n",
    "            fittedModel3 = None\n",
    "            fittedModel4 = None\n",
    "            noLog = True\n",
    "            print('Log failed')\n",
    "        try:\n",
    "            fittedModel5 = smf.ols(formula = formula5, data=data).fit()\n",
    "            fittedModel6 = smf.ols(formula = formula6, data=data).fit()\n",
    "        except:\n",
    "            fittedModel5 = None\n",
    "            fittedModel6 = None\n",
    "            noBox = True\n",
    "            print('Boxcox failed')\n",
    "       \n",
    "        \n",
    "        print()\n",
    "        print('No changes')\n",
    "        print(fittedModel.aic)\n",
    "        print(sm.stats.jarque_bera(fittedModel.resid)[0])\n",
    "        print('Interaction, no transform')\n",
    "        print(fittedModel2.aic)\n",
    "        print('Log transform')\n",
    "        try:\n",
    "            print(fittedModel3.aic)\n",
    "            print(sm.stats.jarque_bera(fittedModel3.resid)[0])\n",
    "            print('Interaction with log transform')\n",
    "            print(fittedModel4.aic)\n",
    "        except AttributeError:\n",
    "            print()\n",
    "        print('Box cox transform')\n",
    "        try:\n",
    "            print(fittedModel5.aic)\n",
    "            print(sm.stats.jarque_bera(fittedModel5.resid)[0])\n",
    "            print('Interaction, box cox transform')\n",
    "            print(fittedModel6.aic)\n",
    "        except AttributeError:\n",
    "            print()\n",
    "        print()\n",
    "        print()\n",
    "        if not noBox and not noLog: #all transforms available\n",
    "            #check which transform is closest to normality\n",
    "            for model in [fittedModel,fittedModel3,fittedModel5]:\n",
    "                if sm.stats.jarque_bera(model.resid)[0] == min([sm.stats.jarque_bera(x.resid)[0] for x in [fittedModel,fittedModel3,fittedModel5]]):\n",
    "                    #check aic to determine if an interaction term should be included\n",
    "                    if model == fittedModel:\n",
    "                        if model.aic < fittedModel2.aic:\n",
    "                            chosenModel = model\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                        elif model.aic - fittedModel2.aic < 2: #aic not significantly different, choose simpler model\n",
    "                            chosenModel = model\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                        else:\n",
    "                            chosenModel = fittedModel2\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                    \n",
    "                    elif model == fittedModel3:\n",
    "                        if model.aic < fittedModel4.aic:\n",
    "                            chosenModel = model\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                        elif model.aic - fittedModel4.aic < 2: #aic not significantly different, choose simpler model\n",
    "                            chosenModel = model\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                        else:\n",
    "                            chosenModel = fittedModel4\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                            \n",
    "                    elif model == fittedModel5:\n",
    "                        if model.aic < fittedModel6.aic:\n",
    "                            chosenModel = model\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                        elif model.aic - fittedModel6.aic < 2: #aic not significantly different, choose simpler model\n",
    "                            chosenModel = model\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                        else:\n",
    "                            chosenModel = fittedModel6\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "            \n",
    "        elif noBox:\n",
    "            for model in [fittedModel,fittedModel3]:\n",
    "                if sm.stats.jarque_bera(model.resid)[0] == min([sm.stats.jarque_bera(x.resid)[0] for x in [fittedModel,fittedModel3]]):\n",
    "                    #check aic to determine if an interaction term should be included\n",
    "                    if model == fittedModel:\n",
    "                        if model.aic < fittedModel2.aic:\n",
    "                            chosenModel = model\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                        elif model.aic - fittedModel2.aic < 2: #aic not significantly different, choose simpler model\n",
    "                            chosenModel = model\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                        else:\n",
    "                            chosenModel = fittedModel2\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                    \n",
    "                    elif model == fittedModel3:\n",
    "                        if model.aic < fittedModel4.aic:\n",
    "                            chosenModel = model\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                        elif model.aic - fittedModel4.aic < 2: #aic not significantly different, choose simpler model\n",
    "                            chosenModel = model\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                        else:\n",
    "                            chosenModel = fittedModel4\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                            \n",
    "\n",
    "        elif noLog:\n",
    "            for model in [fittedModel,fittedModel5]:\n",
    "                if sm.stats.jarque_bera(model.resid)[0] == min([sm.stats.jarque_bera(x.resid)[0] for x in [fittedModel,fittedModel5]]):\n",
    "                    #check aic to determine if an interaction term should be included\n",
    "                    if model == fittedModel:\n",
    "                        if model.aic < fittedModel2.aic:\n",
    "                            chosenModel = model\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                        elif model.aic - fittedModel2.aic < 2: #aic not significantly different, choose simpler model\n",
    "                            chosenModel = model\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                        else:\n",
    "                            chosenModel = fittedModel2\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                    \n",
    "                            \n",
    "                    elif model == fittedModel5:\n",
    "                        if model.aic < fittedModel6.aic:\n",
    "                            chosenModel = model\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                        elif model.aic - fittedModel6.aic < 2: #aic not significantly different, choose simpler model\n",
    "                            chosenModel = model\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "                        else:\n",
    "                            chosenModel = fittedModel6\n",
    "                            modelDict[col] = chosenModel.model.formula\n",
    "        influence = chosenModel.get_influence()\n",
    "        #studentized residuals (deleted residuals)\n",
    "        studentized_residuals = influence.resid_studentized_external\n",
    "        fig,ax = plt.subplots(1,3,figsize=(15,10))\n",
    "        ax[0].plot(chosenModel.fittedvalues,studentized_residuals,'ko',ms=0.5)\n",
    "        ax[0].set_ylabel('Standardised residuals')\n",
    "        ax[0].set_xlabel('Fitted values')\n",
    "        ax[1].hist(studentized_residuals)\n",
    "        ax[1].set_xlabel('Standardised residuals')\n",
    "        sm.graphics.qqplot(studentized_residuals,ax=ax[2])\n",
    "        plt.title(formula)\n",
    "        plt.savefig('REGRESSION_DIAG_PLOTS/'+col+'_oldest_post_model_selection.png',bbox_inches='tight')\n",
    "        plt.show()\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#megaplot of regressions\n",
    "def scale_lightness(rgb, scale_l):\n",
    "    import colorsys\n",
    "    # convert rgb to hls\n",
    "    h, l, s = colorsys.rgb_to_hls(*rgb)\n",
    "    # manipulate h, l, s values and return as rgb\n",
    "    return colorsys.hls_to_rgb(h, min(1, l * scale_l), s = s)\n",
    "def formatPval(pval):\n",
    "#     pval = (mannwhitneyu(list1,list2,alternative='two-sided').pvalue)*correction\n",
    "#     if pval >= 0.05:\n",
    "#         return 'ns'\n",
    "    if pval >= 0.001:\n",
    "        return round(pval,3)\n",
    "    else:\n",
    "        valList = '{:.2e}'.format(pval).split('e')\n",
    "        return valList[0]+'x 10$^{-'+valList[1].lstrip('-').lstrip('0')+'}$'\n",
    "fig, axesList = plt.subplots(10,5,figsize=(16,20),gridspec_kw={'hspace':0.7,'height_ratios':[1,0.25,1,0.07,1,0.07,1,0.25,1,0.07]})\n",
    "axes = []\n",
    "pAxes,lAxes=[],[]\n",
    "for l in axesList[::2]:\n",
    "    pAxes.extend(l)\n",
    "for l in axesList[1::2]:\n",
    "    lAxes.extend(l)\n",
    "featureList = ['gen_len','clen','avg_int','int_count','int_cov',\n",
    "              'CAI','IntDis','gc','gc3','doms',\n",
    "              'uDoms','max_exp','spec','trans_count','motifs',\n",
    "              'PPIs','evolRate','ess','missenseZ','pLI',\n",
    "              'phi','rvis','shet','EvoTol','loftool']\n",
    "namesDict = {'gen_len':('Genomic Length','log(bp)'),'clen':('CDS Length','log(bp)'),'int_count':('Intron Count','count'),\n",
    "             'avg_int':('Mean intron length','log(bp)'),'int_cov':('Intron coverage','%'),'CAI':('Codon optimisation','CAI score'),\n",
    "             'IntDis':('Intrinsic Disorder','%'),'gc':('GC content','%'),'gc3':('GC3 content','%'),\n",
    "             'doms':('Total domains','count'),'uDoms':('Unique domains','count'),'max_exp':('Maximal expression','log(TPM)'),\n",
    "             'spec':('Expression specificity','tau'),'trans_count':('Number of isoforms','count'),\n",
    "             'motifs':('Number of regulatory motifs','count'),'PPIs':('PPIs','count'),'evolRate':('Rate of evolution','dN/dS'),\n",
    "             'ess':('Cellular essentiality','-(CRISPR score)'),'missenseZ':('Missense Z score','score'),'pLI':('pLI','score'),\n",
    "             'phi':('Phi','score'),'shet':('S$_{het}$','score'),'EvoTol':('EvoTol','score'),'loftool':('LoFTool','percentile'),\n",
    "             'rvis':('RVIS','score')}\n",
    "for col, ax, axL in zip(featureList,pAxes,lAxes):\n",
    "    formula=modelDict[col]\n",
    "    data = df\n",
    "    fittedModel = smf.ols(formula = formula, data = data).fit()\n",
    "    valsWGD = df[df['dup_type']=='WGD'][['maxAge',col]]\n",
    "    valsSSD = df[df['dup_type']=='SSD'][['maxAge',col]]\n",
    "    valsSSDLine = df[['maxAge',col]]\n",
    "    valsSSDLine['dup_type'] = 'SSD'\n",
    "    valsWGDLine = df[['maxAge',col]]\n",
    "    valsWGDLine['dup_type'] = 'WGD'\n",
    "    if col in ['evolRate','uDoms','gc','avg_int','ess','missenseZ','shet']: #log transform\n",
    "        print(col)\n",
    "        if min(df[col]) <= 0:\n",
    "            df[col] = [x + (min([y for y in df[col] if y > 0]))/2 for x in df[col]]\n",
    "        valsWGD[col] = np.log10(valsWGD[col])\n",
    "        valsSSD[col] = np.log10(valsSSD[col])\n",
    "    elif col in ['gen_len','clen','gc3','doms','trans_count','int_count','max_exp','int_cov',\n",
    "                 'pLI','PPIs','motifs','rvis','EvoTol','CAI','IntDis']: #box cox transform\n",
    "        if min(df[col]) <= 0:\n",
    "            df[col] = [x + (min([y for y in df[col] if y > 0]))/2 for x in df[col]]\n",
    "        #get lambda for total transform\n",
    "        lam = stats.boxcox(df[col])[1]\n",
    "        valsWGD[col] = stats.boxcox(valsWGD[col],lmbda=lam)\n",
    "        valsSSD[col] = stats.boxcox(valsSSD[col],lmbda=lam)\n",
    "    if col == 'shet':\n",
    "        valsWGD = valsWGD[valsWGD[col]<-0.117]\n",
    "        valsSSD = valsSSD[valsSSD[col]<-0.117]\n",
    "    ax.set_fc((0,0,0,0.05))\n",
    "    ax.grid(color='w',lw=2)\n",
    "#     ax.set_yticks([])\n",
    "    ax.set_xticks([])\n",
    "    ax.plot(valsWGD['maxAge'],valsWGD[col],'o',color=scale_lightness((0.9,0.2,0),1.4),ms=0.7)\n",
    "    ax.plot(valsWGDLine['maxAge'],fittedModel.predict(valsWGDLine),color=scale_lightness((0.9,0.2,0),1.4),lw=2,label='WGD')\n",
    "    ax.plot(valsSSD['maxAge'],valsSSD[col],'o',color=scale_lightness((0.9,0.3,0),1.8),ms=0.7)\n",
    "    ax.plot(valsSSDLine['maxAge'],fittedModel.predict(valsSSDLine),color=scale_lightness((0.9,0.3,0),1.8),lw=2,label='SSD')\n",
    "    ax.set_title(namesDict[col][0])\n",
    "    axL.axis('off')\n",
    "    if col in ['gen_len','clen','evolRate','avg_int','ess','pLI','int_count','int_cov','PPIs','missenseZ']:\n",
    "        axL.text(0.5,2.2,'Age:',ha='center',weight='bold',size=12)\n",
    "        axL.text(0,1.65,'p='+str(formatPval(fittedModel.pvalues['maxAge'])) + ', coeff ='+str(round(fittedModel.params['maxAge'],3)),ha='left',size=12)\n",
    "        axL.text(0.5,1.1,'Duplicate type:',ha='center',weight='bold',size=12)\n",
    "        axL.text(0,0.55,'p='+str(formatPval(fittedModel.pvalues['dup_type[T.WGD]'])) + ', coeff ='+str(round(fittedModel.params['dup_type[T.WGD]'],3)),ha='left',size=12)\n",
    "        if col in ['gen_len','clen','evolRate','avg_int','ess','pLI']:\n",
    "            axL.text(0.5,0,'Age*Duplicate type:',ha='center',weight='bold',size=12)\n",
    "            axL.text(0,-0.55,'p='+str(formatPval(fittedModel.pvalues['maxAge:dup_type[T.WGD]'])) + ', coeff ='+str(round(fittedModel.params['maxAge:dup_type[T.WGD]'],3)),ha='left',size=12)\n",
    "\n",
    "    else:\n",
    "        axL.text(0.5,5,'Age:',ha='center',weight='bold',size=12)\n",
    "        axL.text(0,3,'p='+str(formatPval(fittedModel.pvalues['maxAge'])) + ', coeff ='+str(round(fittedModel.params['maxAge'],3)),ha='left',size=12)\n",
    "        axL.text(0.5,1,'Duplicate type:',ha='center',weight='bold',size=12)\n",
    "        axL.text(0,-1,'p='+str(formatPval(fittedModel.pvalues['dup_type[T.WGD]'])) + ', coeff ='+str(round(fittedModel.params['dup_type[T.WGD]'],3)),ha='left',size=12)\n",
    "    \n",
    "    if col == 'int_cov':\n",
    "        ax.legend(loc=(0.6,0.71))\n",
    "    \n",
    "plt.savefig('totalRegressionModels.png',bbox_inches='tight')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
